example/adversary/adversary_generation.ipynb

```diff
@@ -28,290 +27,312 @@
+    "        acc.update(label, output)\n",
+    "    \n",
+    "    print(\"Train Accuracy: %.2f\\t Train Loss: %.5f\" % (acc.get()[1], train_loss/(i+1)))"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Get pertubation by using fast sign method, check validation change.\n",
-    "See that the validation set was almost entirely correct before the perturbations, but after the perturbations, it is much worse than random guessing."
+    "## Perturbation\n",
+    "\n",
+    "We first run a validation batch and measure the resulting accuracy.\n",
+    "We then perturbate this batch by modifying the input in the opposite direction of the gradient."
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Validation batch accuracy 0.96875\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Get a batch from the testing set\n",
+    "for data, label in test_data:\n",
+    "    data = data.as_in_context(ctx)\n",
+    "    label = label.as_in_context(ctx)\n",
+    "    break\n",
+    "\n",
+    "# Attach gradient to it to get the gradient of the loss with respect to the input\n",
+    "data.attach_grad()\n",
+    "with mx.autograd.record():\n",
+    "    output = net(data)    \n",
+    "    l = loss(output, label)\n",
+    "l.backward()\n",
+    "\n",
+    "acc = mx.metric.Accuracy()\n",
+    "acc.update(label, output)\n",
+    "\n",
+    "print(\"Validation batch accuracy {}\".format(acc.get()[1]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we perturb the input"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
    "metadata": {
     "collapsed": false
    },
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Validation batch accuracy after perturbation 0.40625\n"
+     ]
+    }
+   ],
    "source": [
-    "val_iter.reset()\n",
-    "batch = val_iter.next()\n",
-    "data = batch.data[0]\n",
-    "label = batch.label[0]\n",
-    "arg_map[\"data\"][:] = data\n",
-    "model.forward(is_train=True)\n",
-    "theta = model.outputs[0].asnumpy()\n",
-    "alpha = Softmax(theta)\n",
-    "print(\"Val Batch Accuracy: \", CalAcc(alpha, label.asnumpy()) / batch_size)\n",
-    "#########\n",
-    "grad = LogLossGrad(alpha, label.asnumpy())\n",
-    "out_grad[:] = grad\n",
-    "model.backward([out_grad])\n",
-    "noise = np.sign(data_grad.asnumpy())\n",
-    "arg_map[\"data\"][:] = data.asnumpy() + 0.15 * noise\n",
-    "model.forward(is_train=True)\n",
-    "raw_output = model.outputs[0].asnumpy()\n",
-    "pred = Softmax(raw_output)\n",
-    "print(\"Val Batch Accuracy after pertubation: \", CalAcc(pred, label.asnumpy()) / batch_size)"
+    "data_perturbated = data + 0.15 * mx.nd.sign(data.grad)\n",
```

Comment:
Might be worth bumping up the adjustment, since my first samples were the same as true label.

Reply:
I'd be wary of that because then the images looks a bit crazy, I think maybe I'll just stick a seed in there